src/QldbDriver.ts

```diff
@@ -156,36 +169,95 @@ export class QldbDriver {
-                    throw err;
+    ): Promise<Type> {
+        if (this._isClosed) {
+            throw new DriverClosedError();
+        }
+
+        // Acquire semaphore and get a session from the pool
+        const getSession: (thisDriver: QldbDriver, startNewSession: boolean) => Promise<QldbSession> = 
+            async function(thisDriver: QldbDriver, startNewSession: boolean): Promise<QldbSession> {
+                debug(
+                    `Getting session. Current free session count: ${thisDriver._sessionPool.length}. ` +
+                    `Currently available permit count: ${thisDriver._availablePermits}.`
+                );
+                if (thisDriver._semaphore.tryAcquire()) {
+                    thisDriver._availablePermits--;
+                    try {
+                        let session: QldbSession
+                        if (!startNewSession) {
+                            session = thisDriver._sessionPool.pop();
+                        }
+                        if (startNewSession || session == undefined) {
+                            debug("Creating a new session.");
+                            const communicator: Communicator = 
+                                await Communicator.create(thisDriver._qldbClient, thisDriver._ledgerName);
+                            session = new QldbSession(communicator);
+                        }
+                        return session;
+                    } catch (e) {
+                        // An error when failing to start a new session is always retriable
+                        throw new ExecuteError(e, true, true);
+                    }
+                } else {
+                    throw new SessionPoolEmptyError()
                 }
-                //If it is ISE but not because of transaction expiry, then pick new session and retry the transaction
-                if (isInvalidSessionException(err) && !isTransactionExpiredException(err) ) {
-                    continue;
+            }
+
+        // Release semaphore and if the session is alive return it to the pool and return true
+        const releaseSession: (thisDriver: QldbDriver, session: QldbSession) => boolean = 
+            function(thisDriver: QldbDriver, session: QldbSession): boolean {
+                thisDriver._semaphore.release();
+                thisDriver._availablePermits++;
+                if (session != null && session.isAlive) {
+                    thisDriver._sessionPool.push(session);
+                    return true;
                 } else {
-                    throw err;
+                    return false;
                 }
-            } finally {
-                if (session != null) {
-                    this._returnSessionToPool(session);
+            }
+        
+        retryConfig = (retryConfig == null) ? this._retryConfig : retryConfig;
+        let session: QldbSession;
+        let startNewSession: boolean = true;
```

Comment:
Is this correct?

For the happy case where an execute is successful, the session is placed back into the pool. On another execute, it will create a new session instead of picking up the valid session based on getSession logic

Reply:
was fixed already but I didnt push it up yet